llvm/include/llvm/ADT/ArrayRef.h

```diff
@@ -335,8 +335,9 @@ namespace llvm {
     MutableArrayRef(T *begin, T *end) : ArrayRef<T>(begin, end) {}
 
     /// Construct a MutableArrayRef from a SmallVector.
-    /*implicit*/ MutableArrayRef(SmallVectorImpl<T> &Vec)
-    : ArrayRef<T>(Vec) {}
+    template <typename U>
+    /*implicit*/ MutableArrayRef(const SmallVectorTemplateCommon<T, U> &Vec)
+        : ArrayRef<T>(Vec) {}
```

Comment:
Test coverage? (be good if ADT/Support things were generally committed separately with unit test coverage, tehn used in a subsequent/follow-up patch)

Reply:
Actually I don't need this change anymore, since I moved off of using MutableArrayRef.

Are we writing unit-test for NFC changes? Seems a bit overkill to me.

Reply:
Didn't figure it was NFC - it makes the ctor usable with a wider range of types, so I'd expect the unit test to reflect that wider contract.

And, wait a sec - isn't this invalid? MutableArrayRef needs a non-constant backing store, since it'll expose mutability, but wouldn't `const SmallVectorTemplateCommon<T, U> &` be a constant backing store? So this would allow mutation of a constant sequence?

Reply:
Const seems incorrect here, but this file isn't touched in this revision anymore (in case you missed the revert in the current diff).

> Didn't figure it was NFC - it makes the ctor usable with a wider range of types, so I'd expect the unit test to reflect that wider contract.

We can make something compile that didn't compile before is still NFC to me.
Also we can go all the way with unit-tests, but that's not the historical practice though, there is likely a good middle ground in terms of unit-tests (and historically we've been a bit overly lax on it IMO), but I'm not sure where's the limit here (at some point I'm concerned when the incentives becomes "wrong" and improvements are not done because people prefer to work around instead of jumping through all the hoops).

Reply:
> Const seems incorrect here, but this file isn't touched in this revision anymore (in case you missed the revert in the current diff).

*nod* Caught that, but figured it might be worth pointing out in case it came up again in another patch/later in a series, etc.

> We can make something compile that didn't compile before is still NFC to me.

Not sure I follow that - most changes allow things to compile that didn't before (adding a new function, etc), so I'm not sure I follow that being a relevant criteria. New (albeit fairly simple) functionality was added - ideally an existing unit test that tested the previous version of the function could be slightly modified to test the more general case. (doesn't look like there's existing test coverage, which is unfortunate - though the deduction guide test is probably/maybe close enough to extend & looks like if this ctor was changed the deduction guide should change too - I wouldn't complain too much if the deduction guide test was the only one updated when adding a feature like this, and incidentally tested the ctor too (& to some extent, ensured that the ctors and deduction guides were kept in sync)

Reply:
> Also we can go all the way with unit-tests, but that's not the historical practice though

For the core ADTs we've been pretty good about it, I think. They're generally pretty easy to test (or should be), etc. I get that adding unit test coverage to more complex data structures that aren't so easily isolated isn't as simple.

Reply:
> For the core ADTs we've been pretty good about it, I think

Uh... I would say we've been "OK" at it, recently. And mostly in terms of non-NFC changes (that is: not what I'm doing here).

But if we're looking at the history of ADT, it's far from that, for example what about the testing for: https://github.com/llvm/llvm-project/commit/24f09cc82d3d6f0e0c8c1a342efa6c840f77e324
(that is just a random change I found when looking at the difference between the ArrayRef/MutableArrayRef constructor, ArrayRef being templated, I'm sure I can find plenty more).

Reply:
>  Not sure I follow that - most changes allow things to compile that didn't before (adding a new function, etc), so I'm not sure I follow that being a relevant criteria. 

I feel you're being pedantic about it here: there is no possibe observable behavior different in the compiler behavior, there is code coverage in the project: hence that seems "good enough" for me.
Of course can go the extreme way in terms of testing, but I repeat that this isn't in line with the way we're testing in general: even if ADT has more unit-tests by nature, the "100% coverage through C++ unit-tests" isn't I believe a generalized trade-off we make in LLVM.
Most "NFC" patch are shuffling code around in a way that makes something compile/work that wasn't before, and yet we don't write unit-tests for every single change.

Reply:
Perhaps this is worth a separate discord thread? Not sure.

> Uh... I would say we've been "OK" at it, recently. And mostly in terms of non-NFC changes (that is: not what I'm doing here).

As far as ADT is concerned, I don't think this is NFC. It changes the functionality of the ADT, by allowing new usage that wasn't allowed before. (I think this would be the equivalent of a command line tool that takes ELF files, say, then is generalized to take any object files  -using some existing API that handles all object files - we'd still add a lit test that shows the new functionality - the ability to pass non-ELF Files, in this case at the ADT interface level we'd add test coverage to demonstrate the new functionality, the ability to pass `SmallVectorTemplateCommon`, not just `SmallVectorTemplateCommon`... oh, wait, reading this again - seems like a tradeoff.

The const/non-const buggy part of this, plus the de-generalization (this used to work for any `SmallVectorImpl` which is more general than `SmallVectorTemplateCommon`) - both of those seem worth separate design discussion and test coverage.

Yes, it's pretty simple, but even something this simple has nuance/discussion and I tihnk testing helps demonstrate the intended change in behavior & might've helped diagnose the issues with this change (& avoided it getting lost in another patch).

Reply:
> As far as ADT is concerned, I don't think this is NFC. It changes the functionality of the ADT, by allowing new usage that wasn't allowed before. 

I disagree with your characterization of NFC: it does not add any observable behavior of a correctly running program. This is consistent with a lot of NFC compiler change.

> I think this would be the equivalent of a command line tool that takes ELF files, say, then is generalized to take any object files -using some existing API that handles all object files - we'd still add a lit test that shows the new functionality

We're talking about a container: there is nothing I can do at runtime here that I couldn't do before. This is basically removing a cast at the call-site, it "pleases" the C++ type system: nothing else. And the C++ type system does not persist beyond the parser, so again: no program runtime behavior.
Your example is in a totally different category here.

Reply:
We do test metaprogramming features too. (eg: the ArrayRef tests have a variety of static_asserts to check certain conversions are valid or invalid)

In this case, it actually reduced functionality/narrowed the contract (my mistake in the previous characterization - I'd thought this generalized the behavior) - so if the wider contract were wrong for some reason, I'd love/expected a test to demonstrate that (a static_assert `!is_convertible_v`) and an explanation in a commit message/review why that was the right thing to do... 

(roughly as far as these ADTs is concerned - I think if it's testable behavior that's intended to be part of the public API (not an implementation detail), we should test it - it's easy to do and helpful to keep track of the intended behavior over time - and we do that pretty well these days, partly based on reviews to ensure this, etc)

Perhaps this'd be an easier conversation in person some time. Not sure how to make progress here.

Reply:
I think "NFC" is ambiguous. It certainly should not change the behavior of all accepted input, but the term is ambiguous on whether (a) new symbols can be exported or (b) new uses are allowed.

The latest revision of the pull request does not touch `ArrayRef.h`.
For new functions in ADT, we definitely need a unittest.
For a change of function signature so that more uses are allowed, I think both practices exist.
I try to do this for my own changes for https://maskray.me/blog/2021-08-08-toolchain-testing#the-test-checks-at-the-wrong-layer

Note: the constructor signature change is related to an exported widely-used header and the ctor can be used in many clients.
If it is a change to a type that appears "internal", I'd not bother with a test.

Reply:
> - I think if it's testable behavior that's intended to be part of the public API (not an implementation detail), we should test it - it's easy to do and helpful to keep track of the intended behavior over time

I agree with this characterization: we just disagree on what is  "testable behavior" because I don't see "parser doesn't complain without a 'cast'" as a "behavior of the program".
(of course we want test coverage, that does not necessarily mean individual C++ unit-test of each C++ API, I know you're not saying that right now, but the different isn't super obvious to me).

> Perhaps this'd be an easier conversation in person some time.

I mean we can punt until the next "seemingly NFC" change that touches ADT? :)

Reply:
We can also look at the change I make on the `Value` class: I'm making some API const that weren't const before and vice-versa ; without C++ unit-test to enforce the difference here right now...